analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotTestThisWithIsOperator.cs

```diff
@@ -19,38 +19,111 @@
+        {
+            var switchStatement = (SwitchStatementSyntax)context.Node;
+            if (switchStatement.Expression.RemoveParentheses() is ThisExpressionSyntax
+                && (ContainsTypeCheckInPattern(switchStatement)
+                    || ContainsTypeCheckInCaswSwitchLabel(switchStatement)))
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.Node.GetLocation()));
+            }
+        }
+
+        private static void AnalyzeSwitchExpression(SyntaxNodeAnalysisContext context)
+        {
+            var switchExpression = (SwitchExpressionSyntaxWrapper)context.Node;
+            if (switchExpression.GoverningExpression.RemoveParentheses() is ThisExpressionSyntax
+                && ContainsTypeCheckInPattern(switchExpression))
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.Node.GetLocation()));
+            }
+        }
+
+        private static bool ContainsTypeCheckInCaswSwitchLabel(SwitchStatementSyntax switchStatement) =>
+            switchStatement.Sections.Any(x => x.Labels.Any(x => x is CaseSwitchLabelSyntax caseSwitchLabel
+                                                                && caseSwitchLabel.Value.IsKind(SyntaxKind.IdentifierName)));
+
+        private static bool ContainsTypeCheckInPattern(SyntaxNode syntaxNode) =>
+            syntaxNode.DescendantNodes()
+                      .Where(x => x.IsAnyKind(SyntaxKindEx.ConstantPattern, SyntaxKindEx.DeclarationPattern, SyntaxKindEx.RecursivePattern))
+                      .Any(x => IsTypeCheckOnThis(x));
+
+        private static bool IsTypeCheckOnThis(SyntaxNode pattern)
+        {
+            if (ConstantPatternSyntaxWrapper.IsInstance(pattern)
+                && (ConstantPatternSyntaxWrapper)pattern is var constantPattern)
+            {
+                return constantPattern.Expression.IsKind(SyntaxKind.IdentifierName)
+                       && IsNotInSubPattern(pattern);
+            }
+            else if (DeclarationPatternSyntaxWrapper.IsInstance(pattern))
+            {
+                return IsNotInSubPattern(pattern);
+            }
+            else if (RecursivePatternSyntaxWrapper.IsInstance(pattern)
+                     && (RecursivePatternSyntaxWrapper)pattern is var recursivePattern)
+            {
+                return recursivePattern.Type != null && IsNotInSubPattern(pattern);
+            }
+
+            return false;
+        }
+
+        private static bool IsNotInSubPattern(SyntaxNode node) =>
+            node.FirstAncestorOrSelf<SyntaxNode>(x => x.IsAnyKind(SyntaxKindEx.IsPatternExpression,
+                                                                  SyntaxKindEx.SwitchExpression,
+                                                                  SyntaxKind.SwitchStatement,
+                                                                  SyntaxKindEx.Subpattern)) is var containingNode
```

Comment:
It is puzzling why we need to continue when it returns `null`
```suggestion
                                                                  SyntaxKindEx.Subpattern)) is { } containingNode
```

Reply:
I'm not sure if `IsKind` tolerates null. If so (and I hope so), why do the indirection with variable? Let's just return directly `FirstAncestorOrSelf<>(...).IsKind(...)`